### PR TITLE
feat: implement snapshots as tarballs

### PR DESCRIPTION
**Pull request by wozniakjan**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:
This PR proposes simple support for `VolumeSnapshots` as tarballs created with 
```
tar -C /[src_vol_path] -czvf /[snapshot_path]/[snapshot_name]/[src_vol_name].tar.gz .
```

The reason why `[src_vol_uuid]` is included in the snapshot path is to be able to efficiently discover [`AlreadyExists` cases](https://github.com/kubernetes-csi/csi-test/blob/3fd3e7915d1029e10e81923e4e0cb3de23059c20/pkg/sanity/controller.go#L1470)

It is the second out of three intended PRs to provide basic snapshot workflow including volume cloning from snapshots. It is preceded by https://github.com/kubernetes-csi/csi-driver-nfs/pull/426 which implemented volume cloning from other volumes.

**Which issue(s) this PR fixes**:

Moving forward with snapshots as discussed in https://github.com/kubernetes-csi/csi-driver-nfs/issues/31

**Special notes for your reviewer**:
Some parts of this PR related to helm chart and external-snapshotter are not ideal. As a followup task, I would like to attempt to address https://github.com/kubernetes-csi/external-snapshotter/issues/812. That way all csi-driver helm charts including this one, instead of embedding external-snapshotter CRDs and controller resources and duplicating them, they could just add the official external-snapshotter chart as a dependency. This should also simplify updating to new releases. But not in scope to be addressed here.

**Does this PR introduce a user-facing change?**:

```release-note
feat: volume snapshots
```

### CODE REVIEW

File: deploy/install-driver.sh
Comment (wozniakjan) [inline deploy/install-driver.sh:20]: kind of expecting the next release to be `v4.3.0` to install the external snapshot-controller and CRDs for `master` and `v4.3.0` and higher.

File: pkg/nfs/controllerserver.go
Comment (wozniakjan) [inline pkg/nfs/controllerserver.go:470]: mini cleanup after https://github.com/kubernetes-csi/csi-driver-nfs/pull/426 - calling `err.Error()` was not necessary

File: pkg/nfs/controllerserver.go
Comment (wozniakjan) [inline pkg/nfs/controllerserver.go:489]: mini cleanup after https://github.com/kubernetes-csi/csi-driver-nfs/pull/426 - making the error easier to identify

File: test/external-e2e/testdriver.yaml
Comment (andyzhangx) [inline test/external-e2e/testdriver.yaml:7]: pls also add `snapshotDataSource: true` in `Capabilities`

File: deploy/example/snapshotclass-nfs.yaml
Comment (andyzhangx) [inline deploy/example/snapshotclass-nfs.yaml:9]: I don't think we need to expose these parameters, if a user wants to take a snapshot, these parameters are already defined in original storage class.

File: test/external-e2e/testdriver.yaml
Comment (andyzhangx) [reply test/external-e2e/testdriver.yaml:7]: it should be 
```
SnapshotClass:
    FromName: true
```

**Comment by andyzhangx** [comment]
when the main work of this feature is almost done, pls also add some unit tests, ut coverage is decreasing a lot
```
Overall coverage decreased (-9.8%) to 63.45%
```

File: test/external-e2e/testdriver.yaml
Comment (wozniakjan) [reply test/external-e2e/testdriver.yaml:7]: > pls also add `snapshotDataSource: true` in `Capabilities`

I was hoping to add that in a followup PR but looking at the logs from `pull-csi-driver-nfs-sanity`, I guess it would be easier to include that here.

**Comment by wozniakjan** [comment]
I will turn this to a draft for now and undraft when it's reviewable again

**Comment by wozniakjan** [comment]
fwiw, same result with https://github.com/kubernetes-csi/csi-driver-nfs/pull/430/commits/ae14b0ef07779b46f824dc4c405df24d66580c19 so that is not the culprit. Exploring https://github.com/kubernetes/community/blob/master/contributors/devel/sig-testing/e2e-tests.md now

**Comment by wozniakjan** [comment]
for future reference, `external-e2e` tests don't get external-snapshotter installed by default: https://github.com/kubernetes-csi/csi-driver-nfs/pull/430/commits/0c7e9474774ff1742c67f4b0a69e1960859eae6c

File: deploy/install-driver.sh
Comment (andyzhangx) [inline deploy/install-driver.sh:23]: I think we need to install snapshot controller in charts, also that's configurable

Comment (wozniakjan) [reply deploy/install-driver.sh:23]: good point, I will include that along with the unit test coverage in this PR later this week when time permits. Pitty https://github.com/kubernetes-csi/external-snapshotter/issues/812 is not yet done as that would make the dependency tracking for helm a lot easier.

Comment (andyzhangx) [reply deploy/install-driver.sh:23]: thanks, here is an example of snapshot controller chart config in csi driver charts: https://github.com/kubernetes-sigs/azurefile-csi-driver/blob/8b37cd361590a79f23239cc2fb74f90cf58bd667/charts/latest/azurefile-csi-driver/values.yaml#L131-L155

File: hack/verify-yamllint.sh
Comment (wozniakjan) [inline hack/verify-yamllint.sh:28]: because https://github.com/kubernetes-csi/external-snapshotter/issues/812 is still open, I had a dilemma here. The `install-driver.sh` in https://github.com/kubernetes-csi/csi-driver-nfs/pull/430/commits/d369f94164b8587c0d53d9bce9e2569842924d1c handles external-snapshotter without the requirement to embed and copy/paste third party manifests.

I would like to try to tackle https://github.com/kubernetes-csi/external-snapshotter/issues/812, which would mean csi-driver helm charts will no longer need to embed the manifests and instead could just depend on the official external-snapshotter chart which would remove the need for this arithmetic offset.

File: deploy/install-driver.sh
Comment (wozniakjan) [reply deploy/install-driver.sh:23]: embedded external-snapshotter resources to the helm chart but I am unsure about https://github.com/kubernetes-csi/csi-driver-nfs/pull/430#discussion_r1148641925

**Comment by wozniakjan** [comment]
> when the main work of this feature is almost done, pls also add some unit tests, ut coverage is decreasing a lot
> `Overall coverage decreased (-9.8%) to 63.45%`

@andyzhangx added unit tests covering fundamental cases (happy cases as well as catching most of wrong inputs)
* copy volume from snapshot
* create snapshot of a volume

the coverage decreased still slightly but to a much smaller degree
```
Overall coverage decreased (-1.7%) to 75.405%
```
Are you ok with that or should I attempt to hunt down the remaining uncovered lines?

**Comment by wozniakjan** [comment]
after other PRs merged, there was a go.mod conflict, appended `go mod tidy` to the PR after rebase

**Comment by andyzhangx** [comment]
> after other PRs merged, there was a `go.mod` conflict, appended `go mod tidy` to the PR after rebase

@wozniakjan could you squash the commits to only a few commits?

**Comment by wozniakjan** [comment]
> @wozniakjan could you squash the commits to only a few commits?

Thanks for pinging, rebased and did the best effort to logically squash commits, ptal @andyzhangx

File: charts/latest/csi-driver-nfs/templates/csi-nfs-controller.yaml
Comment (andyzhangx) [inline charts/latest/csi-driver-nfs/templates/csi-nfs-controller.yaml:64]: revert this change

File: charts/latest/csi-driver-nfs/templates/setup-snapshot-controller.yaml
Comment (andyzhangx) [inline charts/latest/csi-driver-nfs/templates/setup-snapshot-controller.yaml:1]: file renaming as 
`setup-snapshot-controller.yaml` --> `csi-snapshot-controller.yaml`

File: deploy/install-driver.sh
Comment (andyzhangx) [reply deploy/install-driver.sh:23]: could you copy the snapshot-controller deployment and CRDs from charts into `deploy` folder?

File: pkg/nfs/controllerserver.go
Comment (wozniakjan) [reply pkg/nfs/controllerserver.go:476]: it's actually already covered in https://github.com/kubernetes-csi/csi-driver-nfs/pull/430/commits/c8c80d7ae77fa23e50843ec1b34320ce0199ee02.

coverage for `TestCopyVolume` contains `copyFromVolume` as well as `copyFromSnapshot` cases.

File: deploy/uninstall-driver.sh
Comment (andyzhangx) [inline deploy/uninstall-driver.sh:42]: shall we use current csi-snapshot-controller.yaml and CRDs to install snapshot controller?

Comment (wozniakjan) [reply deploy/uninstall-driver.sh:42]: tbh, I would love to tackle https://github.com/kubernetes-csi/external-snapshotter/issues/812 next, which would ideally lead to CSI drivers using helm dependencies https://helm.sh/docs/helm/helm_dependency/ and `kubectl kustomize` to install any external-snapshotter resources. This and other CSI drivers, could then track their dependencies on a few lines by referencing upstream sources, instead of copy/pasting external-snapshotter resources in `./deploy` and `./charts`. And this would simplify version bumps for both installation options and keep things up to date, imho.

But I will defer to your expert opinion.

**Comment by smuda** [comment]
Thank you for adding this feature! Will there be a status or event I can listen for to be sure the snapshot has succeeded?

The scenario is shutting down a database pod (to flush to disk), requesting a snapshot and then waiting for it to finish before starting the database pod again.

**Comment by wozniakjan** [comment]
> Will there be a status or event I can listen for to be sure the snapshot has succeeded?

the `VolumeSnapshot` has `status.readyToUse` field which is set to `true` after the snapshot routine has completed
https://github.com/kubernetes-csi/external-snapshotter/blob/129de4cbb3fc5679285d70d191e66976481082a9/client/config/crd/snapshot.storage.k8s.io_volumesnapshots.yaml#L167

And just like you are used to with other kubernetes resources, you can run
```
$ kubectl describe volumesnapshot
```
to debug for any errors

**Review by andyzhangx — APPROVED** [review]
/lgtm

**Pull request merged**
